### PR TITLE
code cleanup: Don't use else after return

### PR DESCRIPTION
**Pull request by qarkai**
Fix clang-tidy [readability-else-after-return](https://releases.llvm.org/11.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/readability-else-after-return.html) warnings. Flattens code a little bit.

### CODE REVIEW

**Comment by ihhub** [comment]
Hi @qarkai , this change is more personal readability change rather than code correction. For example:
```
int getMagicNumber( const bool truth )
{
    if ( truth )
    {
        return 42;
    }
    else {
        return 51;
    }
}
```
With proposed changes the code will looks like:
```
int getMagicNumber( const bool truth )
{
    if ( truth )
    {
        return 42;
    }

    return 51;
}
```

Both versions of the code are eligible and none of them is wrong. For some cases the first version is easier to read, for another - the second.

If we want to have such rule how are we going automatically enforce it for the subsequent changes?

**Comment by qarkai** [comment]
```c++
    if ( truth )
    {
        return 42;
    }

    return 51;
```
We could change it to ` return truth ? 42 : 51;`. I'll rewrite some changes.

> If we want to have such rule how are we going automatically enforce it for the subsequent changes?

I ran tool from IDE but I guess we could check output of something like `clang-tidy --checks=readability-else-after-return -p build **/*.cpp` after running `cmake -B build -DCMAKE_EXPORT_COMPILE_COMMANDS=ON ...`

**Comment by oleg-derevenetz** [comment]
In general, I don't like to enforce such things, because i's a matter of taste. There may be "guard" `if`s, like this:

```
if (check_1) {
    return false;
}

if (check_2) {
    return false;
}

return something();
```

In this case "no else after return" rule is justifiable. But there may be also "choice" `if`s, like this:

```
if (condition) {
    return something();
} else {
    return something_else();
}
```

and in this case "no else after return" rule is not always good in terms of readability. Yes, sometimes it's handy to replace `if` by `?:`, but sometimes it's not:

```
if (condition) {
    do_something();

    return something();
} else {
    do_something_else();

    return something_else();
}
```

**Comment by qarkai** [comment]
@ihhub, @oleg-derevenetz, @idshibanov I'll revert anything you don't like.

File: src/engine/screen.cpp [hunk 1]
```diff
@@ -365,10 +365,7 @@ namespace
 
         bool isVisible() const override
         {
-            if ( _emulation )
-                return fheroes2::Cursor::isVisible();
-            else
-                return fheroes2::Cursor::isVisible() && ( SDL_ShowCursor( SDL_QUERY ) == SDL_ENABLE );
+            return fheroes2::Cursor::isVisible() && ( _emulation || ( SDL_ShowCursor( SDL_QUERY ) == SDL_ENABLE ) );
```
Comment (ihhub) [inline src/engine/screen.cpp:368]: This change compliates the code reading. Could you please revert it?

File: src/fheroes2/battle/battle_cell.h [hunk 2]
```diff
@@ -79,6 +79,8 @@ namespace Battle
         void SetUnit( Unit * );
 
     private:
+        bool isPassable( const Unit & b, int direction ) const;
```
Comment (ihhub) [inline src/fheroes2/battle/battle_cell.h:82]: Could we please rename the first input argument into `monster`? In this case it'll be obvious what **b** is :)

File: src/fheroes2/gui/text.cpp [hunk 3]
```diff
@@ -73,27 +83,20 @@ size_t TextAscii::Size( void ) const
...
-    if ( character < 0x21 || character > fheroes2::AGG::ASCIILastSupportedCharacter( ft ) ) {
-        if ( isSmallFont( ft ) )
-            return 4;
-        else if ( isLargeFont( ft ) )
-            return 12;
-        else
-            return 6;
-    }
-    else {
+    if ( 0x21 <= character && character <= fheroes2::AGG::ASCIILastSupportedCharacter( ft ) )
```
Comment (ihhub) [inline src/fheroes2/gui/text.cpp:86]: Could you please swap arguments for comparison `character >= 0x21`. Reading code in such way is a bit unusual.

File: src/fheroes2/objects/objtown.cpp [hunk 4]
```diff
@@ -48,16 +48,17 @@ int ObjTwba::GetPassable( u32 index0 )
...
-            return DIRECTION_CENTER_ROW | DIRECTION_BOTTOM_ROW | Direction::TOP;
-        else
-            return DIRECTION_CENTER_ROW | DIRECTION_BOTTOM_ROW;
-    }
+
+    int result = DIRECTION_CENTER_ROW | DIRECTION_BOTTOM_ROW;
+
+    // 7, 17, 27, 37, 47, 57, 67, 77
+    if ( index == 7 )
+        result |= Direction::TOP;
```
Comment (ihhub) [inline src/fheroes2/objects/objtown.cpp:59]: We could use like this: `return ( result |  Direction::TOP );`.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @qarkai , I left multiple comments here. Could you please check them?

**Comment by qarkai** [comment]
Unfortunately `clang-tidy` didn't find a bunch of else-after-return code, I'll close this PR then. Sorry for the noise.

**Comment by ihhub** [comment]
Hi @qarkai , no worries about this!

Is it possible to add clang-tidy into our pipeline? As I wrote previously some checks are really useful like unused headers and it'd be great to catch them.

**Comment by qarkai** [comment]
@ihhub I believe it's possible. There are [some actions](https://github.com/marketplace?type=actions&query=clang-tidy+) in marketplace which add support for clang-tidy.
Probably the main question is which checks do you want to run. There are a lot of them: https://clang.llvm.org/extra/clang-tidy/checks/list.html.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub @qarkai Please don't blindly trust to tools such as clang-tidy, especially in terms of "unused" headers. clang-tidy checks only particular build configuration, and it knows nothing about others. For example, on older Xcode versions there was no `abs(int)` in `cmath.h`, which breaks the build:

https://build.macports.org/builders/ports-10.10_x86_64-builder/builds/145739/steps/install-port/logs/stdio

So I added `#include <cstdlib>` to the appropriate files in this commit: https://github.com/ihhub/fheroes2/commit/0998f27b07716055d29ef2a89918cb2879ddcb49, which was subsequently removed in https://github.com/ihhub/fheroes2/commit/9efabb1fe74febdc7ee9297a5a83baea68c70c00 ("Remove unused headers and forward declarations"), and this will break the build again.

**Comment by qarkai** [comment]
Hi @oleg-derevenetz probably those headers should be added under `#ifdef old xcode`. It allows to:
* ignore them for `clang-tidy`, `iwyu`, etc.
* easily remove them when support for old Xcode is dropped

**Comment by oleg-derevenetz** [comment]
> Hi @oleg-derevenetz probably those headers should be added under `#ifdef old xcode`. It allows to:

In this case people who will build this on older Xcode should know somehow that they should add this define via command line, or we should add some code that will recognize older Xcode versions (using CMake, for example) and add this define automatically. Lots of trouble.

**Comment by qarkai** [comment]
At least those includes could be marked with something like `// for old Xcode`.

**Pull request closed without merging**